aten/src/ATen/native/MetaTensor.cpp

```diff
@@ -7,34 +7,29 @@ namespace native {
 // Will be promoted to a public API later, but not now
 Tensor empty_meta(
   IntArrayRef size,
-  const TensorOptions& options_,
-  c10::optional<c10::MemoryFormat> optional_memory_format
+  c10::optional<ScalarType> dtype,
+  c10::optional<Layout> layout,
+  c10::optional<Device> device,
+  c10::optional<bool> pin_memory,
+  c10::optional<c10::MemoryFormat> memory_format
 ) {
-  TORCH_CHECK(
-    !(options_.has_memory_format() && optional_memory_format.has_value()),
-    "Cannot set memory_format both in TensorOptions and explicit argument; please delete "
-    "the redundant setter.");
-  TensorOptions options = options_.merge_in(TensorOptions().memory_format(optional_memory_format));
-
   // TODO: deduplicate this logic with empty_cpu
 
-  auto dtype = options.dtype();
-  auto device = options.device();
   auto tensor = detail::make_tensor<TensorImpl>(
     // NB: We include the computed dispatch key, not because it will actually
     // participate in dispatch, but so that tests like is_sparse/is_cuda
     // give the correct result (a CUDA meta tensor "is cuda").  If we don't
     // like this, remove the computeDispatchKey line
-    DispatchKeySet{DispatchKey::Meta, options.computeDispatchKey()},
-    dtype,
+    DispatchKeySet{DispatchKey::Meta, computeDispatchKey(dtype, layout, device)},
+    dtype.has_value() ? scalarTypeToTypeMeta(*dtype) : get_default_dtype(),
```

Comment:
One thing that is annoying about the change here is it is now possible for `computeDispatchKey` and `dtype.has_value() ? scalarTypeToTypeMeta(*dtype) : get_default_dtype()` to diverg in their behavior (since computeDispatchKey is going to internally recompute the defaulting itself. It would be better if optional dtypes are resolved once up top, then `computeDispatchKey` can make use of the resolved values directly.

Reply:
I switched it to use `dtype_or_default` which is the same defaulting logic used by `computeDispatchKey`.